grype/matcher/apk/matcher.go

```diff
@@ -135,3 +135,81 @@ func vulnerabilitiesByID(vulns []vulnerability.Vulnerability) map[string][]vulne
 
 	return results
 }
+
+func (m *Matcher) findApkPackage(store vulnerability.Provider, d *distro.Distro, p pkg.Package) ([]match.Match, error) {
+	var matches = make([]match.Match, 0)
+
+	// find Alpine SecDB matches for the given package name and version
+	secDbMatches, err := common.FindMatchesByPackageDistro(store, d, p, m.Type())
+	if err != nil {
+		return nil, err
+	}
+
+	cpeMatches, err := m.cpeMatchesWithoutSecDbFixes(store, d, p)
+	if err != nil {
+		return nil, err
+	}
+
+	// keep all secdb matches, as this is an authoritative source
+	matches = append(matches, secDbMatches...)
+
+	// keep only unique CPE matches
+	matches = append(matches, deduplicateMatches(secDbMatches, cpeMatches)...)
+
+	return matches, nil
+}
+
+func (m *Matcher) matchBySourceIndirection(store vulnerability.Provider, d *distro.Distro, p pkg.Package) ([]match.Match, error) {
+	metadata, ok := p.Metadata.(pkg.ApkMetadata)
+	if !ok {
+		return nil, nil
+	}
+
+	// ignore packages without source indirection hints or where source name is identical to package name
+	if metadata.OriginPackage == "" || metadata.OriginPackage == p.Name {
+		return []match.Match{}, nil
+	}
+
+	// use source package name for exact package name matching
+	var indirectPackage pkg.Package
+
+	err := copier.Copy(&indirectPackage, p)
+	if err != nil {
+		return nil, fmt.Errorf("failed to copy package: %w", err)
+	}
+
+	// use the source package name
+	indirectPackage.Name = metadata.OriginPackage
+
+	//For each cpe, replace pkg name with origin and add to set
+	CPEStrings := strset.New()
+	for _, cpe := range indirectPackage.CPEs {
+		updatedCPEString := strings.Replace(cpe.BindToFmtString(), p.Name, indirectPackage.Name, -1)
+		CPEStrings.Add(updatedCPEString)
+	}
+
+	//With each entry in set, convert string to CPE and update indirectPackage CPEs
+	updatedCPEs := make([]syftPkg.CPE, 0)
+	for _, CPEString := range CPEStrings.List() {
+		updatedCPE, _ := syftPkg.NewCPE(CPEString)
+		updatedCPEs = append(updatedCPEs, updatedCPE)
+	}
+	indirectPackage.CPEs = updatedCPEs
+
+	matches, err := m.findApkPackage(store, d, indirectPackage)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find vulnerabilities by apk source indirection: %w", err)
+	}
+
+	// we want to make certain that we are tracking the match based on the package from the SBOM (not the indirect package)
+	// however, we also want to keep the indirect package around for future reference
+	for idx := range matches {
+		matches[idx].Package = p
+
+		if matches[idx].Type == match.ExactDirectMatch {
```

Comment:
Is it important that we perform this check first before setting `matches[idx].Type` to `match.ExactIndirectMatch`? If so, why?

Reply:
I believe that this check needs to be done because the matches slice contains findings against both secdb and nvd for source package. The ones found against secdb for source package should be changed to ExactIndirectMatch, while the ones found against nvd for source pacakge should remain FuzzyMatches. There is no type that reflect a fuzzy match made against source package so @wagoodman suggested I make https://github.com/anchore/grype/issues/408 to reflect this.

Reply:
Ahhhh — that makes complete sense. I forgot about the "fuzzy" value that's possible here, and it sounds like this is a good solution given the non-ideal state of the pre-#408 world.